ChatSecure/Classes/View Controllers/OTRMessagesViewController+SupplementaryViews.swift

```diff
@@ -105,21 +119,36 @@ import OTRAssets
     open func removeSupplementaryViewsOfType(type:String) {
         for (key,value) in self.supplementaryViews {
             var newValue = value
-            if let index = newValue.index(of: type) {
+            if let index = newValue.index(where: { (viewKind,userData) -> Bool in
+                return viewKind == type
+            }) {
                 newValue.remove(at: index)
                 supplementaryViews[key] = newValue
             }
         }
     }
-    
-    open func addSupplementaryViewForMessage(message:OTRMessageProtocol, supplementaryView type:String) {
-        var value = self.supplementaryViews[message.uniqueId]
-        if value == nil {
-            value = [type]
+
+    open func removeSupplementaryView(indexPath:IndexPath, supplementaryView type:String) {
+        var value = self.supplementaryViews[indexPath] ?? []
+        if let index = value.index(where: { (viewKind, _) -> Bool in
+            return viewKind == type
+        }) {
+            value.remove(at: index)
+            self.supplementaryViews[indexPath] = value
+        }
```

Comment:
This pattern is repeated in a number of places and should be factored out.

```swift
var value = self.supplementaryViews[indexPath] ?? []
 if let index = value.index(where: { (viewKind, _) -> Bool in
  return viewKind == type
   }) {
```

I think it may be possible to simplify this using another data structure.